### PR TITLE
Add Output File

### PR DESCRIPTION
**Pull request by ScruffyProdigy**

### CODE REVIEW

File: google/internal/externalaccount/executablecredsource.go [hunk 1]
```diff
@@ -22,34 +24,43 @@ const (
...
+	return fmt.Errorf("oauth2/google: %v contains unsuccessful response: (%v) %v", source, code, message)
 }
 
-func unsupportedVersionError(version int) error {
-	return fmt.Errorf("oauth2/google: executable returned unsupported version: %v", version)
+func unsupportedVersionError(source string, version int) error {
+	return fmt.Errorf("oauth2/google: %v contains unsupported version: %v", source, version)
+}
+
+type timeoutException struct {
```
Comment (codyoss) [inline google/internal/externalaccount/executablecredsource.go:51]: Any reason for the struct here?

File: google/internal/externalaccount/executablecredsource.go [hunk 2]
```diff
@@ -180,32 +191,61 @@ func parseSubjectToken(response []byte) (string, error) {
...
+	}
+
+	data, err := ioutil.ReadAll(io.LimitReader(file, 1<<20))
+	if err != nil {
+		// An error reading the file. Not necessarily under the developer's control, so ignore it
+		return "", nil, false
+	}
+
+	token, err := parseSubjectTokenFromSource(data, outputFileSource)
+
```
Comment (codyoss) [inline google/internal/externalaccount/executablecredsource.go:236]: nit: remove space.

File: google/internal/externalaccount/executablecredsource.go [hunk 3]
```diff
@@ -180,32 +191,61 @@ func parseSubjectToken(response []byte) (string, error) {
...
+	if err == nil {
```
Comment (codyoss) [inline google/internal/externalaccount/executablecredsource.go:237]: try to flip this logic to happy path is returned at the end. Indented path is usually the unhappy path in Go

File: google/internal/externalaccount/executablecredsource.go (hunk 1 above)
Comment (ScruffyProdigy) [reply google/internal/externalaccount/executablecredsource.go]: I've renamed this to nonCacheableError.  Some errors aren't errors when they're in the output file, they just mean that we need to check the executable, and I'm using this error type to indicate that

File: google/internal/externalaccount/executablecredsource.go [hunk 4]
```diff
@@ -180,32 +191,61 @@ func parseSubjectToken(response []byte) (string, error) {
...
-func (cs executableCredentialSource) getTokenFromOutputFile() (string, bool) {
-	// TODO
-	return "", false
+func (cs executableCredentialSource) getTokenFromOutputFile() (string, error, bool) {
+	if cs.OutputFile == "" {
+		// This ExecutableCredentialSource doesn't use an OutputFile
+		return "", nil, false
+	}
+
+	file, err := os.Open(cs.OutputFile)
```
[the remark this answers is not in the document]
Comment (ScruffyProdigy) [reply google/internal/externalaccount/executablecredsource.go:224]: Done

File: google/internal/externalaccount/executablecredsource.go (hunk 2 above)
Comment (ScruffyProdigy) [reply google/internal/externalaccount/executablecredsource.go:244]: Done

File: google/internal/externalaccount/executablecredsource.go (hunk 3 above)
Comment (ScruffyProdigy) [reply google/internal/externalaccount/executablecredsource.go]: Done

**Review by codyoss — APPROVED** [review]

File: google/internal/externalaccount/executablecredsource.go [hunk 5]
```diff
@@ -180,32 +192,61 @@ func parseSubjectToken(response []byte) (string, error) {
...
 
 	return cs.getTokenFromExecutableCommand()
 }
 
-func (cs executableCredentialSource) getTokenFromOutputFile() (string, bool) {
-	// TODO
-	return "", false
+func (cs executableCredentialSource) getTokenFromOutputFile() (string, error, bool) {
+	if cs.OutputFile == "" {
+		// This ExecutableCredentialSource doesn't use an OutputFile
```
Comment (lsirac) [inline google/internal/externalaccount/executablecredsource.go:220]: missing period

**Review by lsirac — APPROVED** [review]

File: google/internal/externalaccount/executablecredsource.go (hunk 5 above)
Comment (ScruffyProdigy) [reply google/internal/externalaccount/executablecredsource.go]: Done

**Pull request merged**